### DISCUSSION TITLE
GraalVM

### DISCUSSION THREAD

**Discussion by cemo** [open-telemetry/opentelemetry-java-instrumentation#6908] [Q&A]
Is there a way to use OpenTelemetry without instrumentation? I would like to use OpenTelemetry by Graal but I have no idea how to solve this problem.

**Comment by laurit** [comment]
https://github.com/open-telemetry/opentelemetry-java provides apis for manual instrumentation.

**Reply by trask** [reply]
also, long-term we are hoping to have a solution for applying auto-instrumentation during the native compilation process (see early work towards this solution at https://github.com/open-telemetry/opentelemetry-java-contrib/tree/main/static-instrumenter)

**Comment by cemo** [comment]
@trask Thank you for your answer. This is the thing I am looking for. Is there a road map about it? It sounds really promising.

**Comment by cemo** [comment]
@jeanbisutti  is it possible to use along with [Google Jib project](https://github.com/GoogleContainerTools/jib)?

**Comment by jeanbisutti** [comment]
@cemo It's not possible today to use the AOT/static instrumentation with Graal. We have identified several issues. I have noticed problems with Spring Boot and the static instrumentation code. I will create PRs to fix them. There are also OTel initialization issues, like this [one](https://github.com/open-telemetry/opentelemetry-java-contrib/issues/506). To avoid them, I am trying to have a Java agent (with OTel initializations and the JDK instrumentation) and the instrumentation, other than the JDK ones, done AOT. Except for Graal, this approach may allow starting to have something working with AOT instrumentation and improve it. This approach has to be discussed on the OTel side and might help some applications with a limited number of cores to get started faster. The next step may be to do the OTel initializations without an agent and have something working with Graal (and without the instrumentation of JDK classes).

**Comment by cemo** [comment]
Hey @trask @jeanbisutti, Is there an improvement on this issue?

There is an ongoing afford on this PR and waiting to be reviewed. Can someone take a look please as well?

https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/11068

**Reply by steverao** [reply]
Yes, this is an improvement to achieve auto-instrumentation under GraalVM. So far, @ziyilin and I tested some plugins and found they can run well based on the approach. But there are some problems that be mentioned in Limitations  part of the PR need to solve in the future, if we want to have exactly the same effect as non-GraalVM.

**Comment by jeanbisutti** [comment]
@cemo For Spring Boot you can use the [OpenTelemetry starter](https://opentelemetry.io/docs/languages/java/automatic/spring-boot/) that works with GraalVM native.
